### ISSUE TITLE
[question] Define aliases that depends on settings and options

### ISSUE THREAD

**Issue by lucabonifacio** [conan-io/conan#9510]
It possible to define alias packages that use the settings and options to define the target packages?

My first approach, would be, for being able to get access to the `settings` and `options` when Conan is loading the recipe, to re-define the attribute `alias` as a property using python decorator `@property`. By doing this, the `self` context is accessible. Here the recipe for the alias.
```
from conans import ConanFile
from conans.errors import ConanInvalidConfiguration

class BoostAliasConan(ConanFile):
    name = 'boost'
    version = 'latest'   
    revision_mode = 'hash'

    settings = 'os', 'compiler'

    options = {
        'shared': [True, False],
    }
    default_options = {
        'shared': True,
    }

    @property
    def alias(self):
        if self.settings.os == 'Windows':
            if self.options.shared:
                boost_version = '1.57.0'
            else:
                boost_version = '1.63.0'
        elif self.settings.os == 'Linux':
            boost_version = '1.72.0'
        else:
            raise ConanInvalidConfiguration('This configuration is not supported')

        return f'boost/{boost_version}@user/stable'
```

When using the alias package `boost/latest@user/stable` without specifying any option, the alias resolution is fine. But specifying option `shared=False` on `install`, then the alias recipe still will have the default value set. this results in getting boost version `1.57.0` instead of `1.63.0`.

What would be the right way to define such aliases?

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Hi @lucabonifacio 

No, it is not possible. Alias recipes are not evaluated, they do not get settings and options, they are to be used as a "static" pointer, not a dynamic one. Are you saying that the above is working for ``self.settings.os`` but not for options? I don't think it does.

Alias recipes is a tricky feature, we recently had to introduce a new syntax to help resolving them, because they were having too many issues related their resolution. I really think that we don't want to add extra complexity into them, and allowing them to provide dynamic resolution based on settings and options, but I'll ask for further feedback.

**Comment by lasote** [comment]
I think it would be pretty complex for the value.

**Comment by dornbirndevelops** [comment]
> No, it is not possible. Alias recipes are not evaluated, they do not get settings and options, they are to be used as a "static" pointer, not a dynamic one. Are you saying that the above is working for `self.settings.os` but not for options? I don't think it does.

Hi all!

I can confirm that an alias recipe like above does retrieve the settings context but not the options context.
Also, it can be seen that the alias is evaluated twice (the log entry appears twice).
Tested on version 1.39.0

alias/conanfile.py
```python
from conans import ConanFile

class FooConan(ConanFile):
    name = "foo"
    version = "latest"
    
    settings = "os"
    options = { "shared": [True, False] }
    default_options = { "shared": False }

    revision_mode = "hash"

    @property
    def alias(self):
        print(f"os: {self.settings.os}")
        print(f"shared: {self.options.shared}")
        return 'foo/1.0@'
```

`conan export alias`
...

`conan install foo/latest@ -s os=Windows -o *:shared=True`
```
Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=Visual Studio
compiler.runtime=MD
compiler.version=16
os=Windows
os_build=Windows
[options]
*:shared=True
[build_requires]
[env]

os: Windows
shared: False
os: Windows
shared: False
foo/1.0: Not found in local cache, looking in remotes...
foo/1.0: Trying with 'conan-center'...
foo/1.0: Trying with 'leica-sandbox'...
ERROR: Unable to find 'foo/1.0' in remotes
```

`conan install foo/latest@ -s os=Linux -o *:shared=True`
```
Configuration:
[settings]
arch=x86_64
arch_build=x86_64     
build_type=Release    
compiler=Visual Studio
compiler.runtime=MD   
compiler.version=16   
os=Linux
os_build=Windows      
[options]
*:shared=True
[build_requires]      
[env]

os: Linux
shared: False
os: Linux
shared: False
foo/1.0: Not found in local cache, looking in remotes...
foo/1.0: Trying with 'conan-center'...
foo/1.0: Trying with 'leica-sandbox'...
ERROR: Unable to find 'foo/1.0' in remotes
```

**Comment by memsharded** [comment]
I am afraid that such is some bad behavior, possibly because some internal caching in the legacy ``requires = "foo/latest"`` syntax. The new explicit alias syntax approved by the Tribe for 2.0 (and already available at least since 1.39), will fail exactly in this step, because the settings and options are not initialized, this test proves it:

```python
def test_alias_settings():
    c = TestClient()
    c.save({"conanfile.py": GenConanfile()})
    c.run("create . foo/1.0@")
    c.save({"conanfile.py": textwrap.dedent("""
        from conans import ConanFile

        class FooConan(ConanFile):
            settings = "os"
            options = { "shared": [True, False] }
            default_options = { "shared": False }

            @property
            def alias(self):
                self.output.info("os: {}".format(self.settings.os))
                print("shared: {}".format(self.options.shared))
                return 'foo/1.0@'
        """)})
    c.run("export . foo/latest@")

    # Legacy syntax, might load settings/options, incorrectly
    c.save({"conanfile.py": GenConanfile().with_requires("foo/latest")})
    c.run("install .")
    # New syntax will raise
    c.save({"conanfile.py": GenConanfile().with_requires("foo/(latest)")})
    c.run("install .", assert_error=True)
    assert "Alias definition error in foo/latest" in c.out
```

**Comment by memsharded** [comment]
To follow up: it is not possible to use settings and options in alias to decide dynamically, and it would be quite complicated to implement, adding too much extra complexity when we are trying to simplify the alias logic.

In any case, it is possible to define a "proxy" recipe that implements the logic, (and can remove the variance of packages removing settings and options in ``package_id()``.) That would be a regular recipe, not an alias, and will be represented in the dependency graph, but I don't see a reason this won't work.

I'd say the question is responded and can be closed, if that approach doesn't work, we could discuss why and what are the alternatives, but at the moment I am closing this because because it doesn't seem something possible to implement reasonably.
